hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DistributedFileSystem.java

```diff
@@ -2901,6 +2945,74 @@ private void provisionEZTrash(String path, FsPermission trashPermission)
     setPermission(trashPath, trashPermission);
   }
 
+  /**
+   * HDFS only.
+   * 
+   * Provision snapshottable directory trash.
+   * @param path Path to a snapshottable directory.
+   * @param trashPermission Expected FsPermission of the trash root.
+   * @throws IOException
+   */
+  public void provisionSnapshottableDirTrash(final Path path,
+      final FsPermission trashPermission) throws IOException {
+    Path absF = fixRelativePart(path);
+    new FileSystemLinkResolver<Void>() {
+      @Override
+      public Void doCall(Path p) throws IOException {
+        provisionSnapshottableDirTrash(getPathName(p), trashPermission);
+        return null;
+      }
+
+      @Override
+      public Void next(FileSystem fs, Path p) throws IOException {
+        if (fs instanceof DistributedFileSystem) {
+          DistributedFileSystem myDfs = (DistributedFileSystem)fs;
+          myDfs.provisionSnapshottableDirTrash(p, trashPermission);
+          return null;
+        }
+        throw new UnsupportedOperationException(
+            "Cannot provisionSnapshottableDirTrash through a symlink to" +
+            " a non-DistributedFileSystem: " + fs + " -> " + p);
+      }
+    }.resolve(this, absF);
+  }
+
+  private void provisionSnapshottableDirTrash(
+      String pathStr, FsPermission trashPermission) throws IOException {
+    Path path = new Path(pathStr);
```

Comment:
Hi Nicholas, thanks for reviewing the patch.

I thought about placing the check, but I eventually didn't. Only admins can use it since this is a `dfsadmin` command. Maybe the admin just want to provision snapshot trash before restarting the NameNode to enable snapshot trash root.

That being said, I want to add a warning if `isSnapshotTrashRootEnabled == false` when running this command. So the admin can be reminded. What do you think?

Reply:
Sure, adding a warning sounds good.